### PR TITLE
Add links to external classes in clients javadoc

### PR DESCRIPTION
**Pull request by javanna**
The client sniffer depends on the low-level REST client, while the Java high-level REST client and the transport client depend on Elasticsearch itself. Javadoc are not that useful unless they have links to the Elasticsearch classes in the latter case, and to the low-level REST client in the sniffer javadoc. This commit adds those links.

### CODE REVIEW

File: client/rest-high-level/build.gradle [hunk 1]
```diff
@@ -59,4 +59,17 @@ forbiddenApisMain {
...
+
+javadoc {
+    dependsOn ':core:javadoc'
+    dependsOn ':client:rest:javadoc'
+    String host;
+    if (version.endsWith("-SNAPSHOT")) {
+        host = "https://snapshots.elastic.co";
+    } else {
+        host = "https://artifacts.elastic.co";
+    }
```
Comment (javanna) [inline client/rest-high-level/build.gradle]: suggestions are welcome on where this repeated if can be moved and shared across the three projects that need it.

**Comment by javanna** [comment]
I tried linking the http client etc. javadoc as well, which would be nice now that such dependencies are shaded, but I had no luck with that.

File: client/rest-high-level/build.gradle (hunk 1 above)
Comment (tlrx) [inline client/rest-high-level/build.gradle]: Maybe we could use something like `ext.javadoc_host` (like `ext.bwc_tests_enabled` in elasticsearch build.gradle file) and reuse it in the sub projects.

**Review by tlrx — COMMENTED** [review]
That would be nice to find a way to mutualize the host & condition on version.

> I tried linking the http client etc. javadoc as well, which would be nice now that such dependencies are shaded, but I had no luck with that.

The only way I know is to post-process the generated javadoc files to change the HTML links.

File: client/transport/build.gradle [hunk 2]
```diff
@@ -55,3 +55,15 @@ namingConventions {
...
+
+javadoc {
+  dependsOn ':core:javadoc'
+  String host;
+  if (version.endsWith("-SNAPSHOT")) {
+    host = "https://snapshots.elastic.co";
+  } else {
+    host = "https://artifacts.elastic.co";
+  }
+  options.linksOffline host + "/javadoc/org/elasticsearch/elasticsearch/${version}", "../../core/build/docs/javadoc/"
```
Comment (rjernst) [inline client/transport/build.gradle]: I think this could be done generically. So inside BuildPlugin, have an iteration over all the projects that depend on the current project, add the javadoc dependsOn, and the linksOffline then for the locally build javadocs. Then any gradle project that eg depends on core automatically can link to its javadocs.

Comment (nik9000) [inline client/transport/build.gradle]: Would you do it for all dependencies or just core?

File: client/sniffer/build.gradle [hunk 3]
```diff
@@ -100,3 +100,14 @@ thirdPartyAudit.excludes = [
   'javax.servlet.ServletContextEvent',
   'javax.servlet.ServletContextListener'
 ]
+
+javadoc {
+  dependsOn ':client:rest:javadoc'
```
Comment (javanna) [inline client/sniffer/build.gradle]: I think they can be built in any order, but if you do gradle :client:sniffer:javadoc it will run :client:rest:javadoc before that?

File: client/transport/build.gradle (hunk 2 above)
Comment (javanna) [inline client/transport/build.gradle]: it is appealing to do it generically, yet I think things will quickly become complicated. For instance the order of the linksOffline options (rest and core) in rest-high-level matters, as we want to do first `org.elasticsearch.client.*` and then `org.elasticsearch.*`, otherwise the links to the RestClient classes point to core instead (non existing html files). I think doing it only for elasticsearch deps may work, but given that these are the only projects that we promote the javadocs for at the moment (links are in the docs pages), I don't see a big advantage in doing it generically at the moment.

Comment (rjernst) [inline client/transport/build.gradle]: The order is already defined. Javadocs of dependencies must be built before the current project's javadocs.

Comment (javanna) [inline client/transport/build.gradle]: I don't mean order of dependencies. I mean order of linksOffline options. In my experience that counts, no matter if the deps were properly built. If org.elasticsearch.* is first, the links to the low-level client won't work. At least that is what I saw.

**Comment by javanna** [comment]
I pushed an update. The javadoc links to classes belonging to artifacts whose group starts with "org.elasticsearch" are added automatically. That works for modules, plugins, rest high-level client, test framework etc. @rjernst hopefully this is what you had in mind, can you have another look please?

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy [hunk 4]
```diff
@@ -79,6 +78,30 @@ class BuildPlugin implements Plugin<Project> {
         configureConfigurations(project)
         project.ext.versions = VersionProperties.versions
         configureCompile(project)
+
+        String artifactsHost = project.ext.versions.elasticsearch.endsWith("-SNAPSHOT") ? "https://snapshots.elastic.co" : "https://artifacts.elastic.co"
```
Comment (tlrx) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy]: Maybe this could go in a `configureJavadoc` method?

**Review by tlrx — APPROVED** [review]
LGTM, I left a comment

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy [hunk 5]
```diff
@@ -79,6 +78,30 @@ class BuildPlugin implements Plugin<Project> {
...
+             * or all the links for the other packages (e.g org.elasticsearch.client) will point to core rather than their own artifacts
+             */
+            Closure sortClosure = { a, b -> b.group <=> a.group };
+            Closure depJavadocClosure = { dep ->
+                if (dep.group != null && dep.group.startsWith('org.elasticsearch')) {
+                    String substitution = project.ext.projectSubstitutions.get("${dep.group}:${dep.name}:${dep.version}")
+                    if (substitution != null) {
+                        project.javadoc.dependsOn substitution + ':javadoc'
+                        String artifactPath = dep.group.replaceAll('\\.', '/') + '/' + dep.name.replaceAll('\\.', '/') + '/' + dep.version
+                        String projectPath = substitution.replaceAll(':', '/')
```
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy]: This isn't always true. You should use `project(substitution).path`.  But I don't think you need this at all. Instead of `"${project.rootDir}${projectPath}/build/docs/javadoc/"` below, use `"${project(substitution).buildDir}/docs/javadoc/"`

File: test/framework/build.gradle [hunk 6]
```diff
@@ -27,7 +27,6 @@ dependencies {
   compile "org.hamcrest:hamcrest-all:${versions.hamcrest}"
   compile "org.apache.lucene:lucene-test-framework:${versions.lucene}"
   compile "org.apache.lucene:lucene-codecs:${versions.lucene}"
-  compile "org.elasticsearch.client:elasticsearch-rest-client:${version}"
```
Comment (rjernst) [inline test/framework/build.gradle]: Why was this removed? Isn't it needed by the rest test runner?

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy (hunk 4 above)
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy]: You can also just look at `VersionProperties.elasticsearch`, instead of `project.ext.versions.elasticsearch`

**Review by rjernst — APPROVED** [review]
Looks ok. I'm not sure using Configuration.dependencies will work long term (it only has the direct dependencies), but I guess this is ok for now.  Please do double check that that is not causing a dependency resolution during configuration (run a simple command like `gradle -i`, and see if anything indicates any dependencies are being resolved).

File: test/framework/build.gradle (hunk 6 above)
Comment (javanna) [inline test/framework/build.gradle]: it was duplicated

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy (hunk 5 above)
Comment (javanna) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy]: good one, thanks
